### PR TITLE
Rule matches optimization: Fix for #183

### PR DESCRIPTION
**Pull request by piyushroshan**
Fix for #183 

> Thank you for contributing to Coraza WAF, your effort is greatly appreciated
> Before submitting check if what you want to add to `coraza` list meets [quality standards](https://github.com/corazawaf/coraza/blob/master/CONTRIBUTING.md#quality-standards) before sending pull request. Thanks!

**Note**: _that go.mod and go.sum can only be modified for tested dependency updates or justified new features._

**Make sure that you've checked the boxes below before you submit PR:**

- [x] My code includes positive and negative tests.
- [x] I have an appropriate description with correct grammar.
- [x] I have read [Contribution guidelines](https://github.com/corazawaf/coraza/blob/master/CONTRIBUTING.md#contribution-guidelines), [maintainers note](https://github.com/corazawaf/coraza/blob/master/CONTRIBUTING.md#maintainers) and [Quality standard](https://github.com/github.com/corazawaf/coraza/v2sso/coraza-waf/blob/master/CONTRIBUTING.md#quality-standards).

Thanks for your PR :heart:

### CODE REVIEW

**Review by fzipi — APPROVED** [review]

**Comment by jptosso** [comment]
LGTM, note we will have to force merge after the second approval because of a bug in the regression tests.

Thank you for your contribution

**Review by jptosso — APPROVED** [review]

File: rule_match.go
Comment (fzipi) [inline rule_match.go:103]: While I see that this was here before, can we switch this 200 to a package const?

Maybe downstream someone wants to log more than that, so it would be awesome if it can be configured...

**Review by fzipi — COMMENTED** [review]
Added one comment

File: rule.go
Comment (bxlxx) [inline rule.go:356]: Maybe it's more suitable to use ```matchedChainValues := []MatchData{}``` here.

File: rule.go
Comment (bxlxx) [inline rule.go:366]: There seem to be some problems here. You didn't add the ```matchedData``` to the ```matchedChainValues```  in the for loop. Data will be lost when merging here.

File: rule_match.go
Comment (bxlxx) [inline rule_match.go:66]: If there is no place to use, you can delete it, thanks.

File: rule.go
Comment (bxlxx) [inline rule.go:356]: Maybe ```matchedChainValues := []MatchData{}``` is more appropriate.

**Review by bxlxx — CHANGES_REQUESTED** [review]

File: rule.go
Comment (piyushroshan) [reply rule.go]: Have moved this in inner loop to preserve intermediate matchedChainValues

File: rule.go
Comment (piyushroshan) [reply rule.go]: Fixed it.

File: rule_match.go
Comment (piyushroshan) [reply rule_match.go:103]: Will make such change in another PR. There are various places this 200 limit is used.

File: rule_match.go
Comment (piyushroshan) [reply rule_match.go:66]: Juan advised against it. As someone already using it could cause errors.
We can remove it in next release.

Comment (jptosso) [reply rule_match.go:66]: @bxlxx we should mark unused exported variables as deprecated before removing them in the next major release.

**Comment by jptosso** [comment]
I'm good with the changes, future improvements like constants should be part of a project, maybe we could create a package like types for configurations. Or keep them on each respective package. 

```go
// constants/constants.go
const ErrorLogMessageLength = 200
/// ...

**Review by jptosso — APPROVED** [review]

**Review by fzipi — APPROVED** [review]
Fine by me. Let's create a ticket s we don't forget about constants.

**Review by bxlxx — APPROVED** [review]
LGTM

**Pull request merged**
